### ISSUE TITLE
Resolve CoralNet and Spacer 0 vs 1 indexing misalign.

### ISSUE THREAD

**Issue by beijbom** [beijbom/coralnet#314]
Unfortunately, we have a situation where CoralNet uses 1-indexed `row, col` locations while spacer uses 0-indexed. This thread is to discuss ways to mitigate. 

First: why is this a problem? Well, the new spacer is much better at tracking `row, col` locations throughout the process, and I was meaning to double check against that at every step when parsing return messages and data structures. We also store `row, col` in the actual `.json` feature files which are part of the CoralNet s3 blob store. 

So what can we do? Well, there are three options.

**1) write converters:** The most obvious option is to write converters at every interaction point between spacer and Coralnet. These include at least:
- Submitting features for extraction.
- Reading feature extraction return.
- Submitting a new training job.
- Submitting and parsing classification jobs.

Also: the [stored features](https://github.com/beijbom/pyspacer/blob/master/spacer/data_classes.py#L142) for an image would use zero-indexing. So it's not just the messaging code that needs to be aware of this, we would be creating a permanent records using two different conventions.

**2) Change spacer to use 1-indexing:** This is probably the easiest to achieve, by, for example subtracting 1 from center [here](https://github.com/beijbom/pyspacer/blob/5158bf1e6f116fe70ff06ed71705326f73c44dfe/spacer/extract_features_utils.py#L54), and also update the [check_rowcols](https://github.com/beijbom/pyspacer/blob/5158bf1e6f116fe70ff06ed71705326f73c44dfe/spacer/task_utils.py#L6) method.

What I don't like about this is that 0-indexing is the "correct" indexing, in the sense that is it used by most (all) python libraries such as PIL,  numpy, in addition to python itself.

**3) Change CoralNet to 0-indexing:** This is probably more work and would include
- database migration. This one seems straight-forwards since it just one table.
- Update code that generates row/cols to use 0-indexing. E.g. [here](https://github.com/beijbom/coralnet/blob/master/project/annotations/model_utils.py#L96
).
- Update consumer code, such as renderings and cropping to use 0-indexing.
- Update import from CPCee and in general make it clear for consumers that we use 0-indexing. (Well, we could of-course use 0-indexing internally but 1-indexing when exporting, but that might be another invitation for trouble)

On the "Update consumer code" topic, I looked around a bit, and it seems some of the  consumer code is already 0-indexed, notably 
1)  crop-generation [code](https://github.com/beijbom/coralnet/blob/master/project/visualization/utils.py#L129
).
2) I'm not an expert on java-script, but since I'd assume java-script uses 0-indexing and I didn't see any `- 1`s I wonder if that one is already using 0-indexing? Also, lines like [this](https://github.com/beijbom/coralnet/blob/master/project/annotations/static/js/AnnotationToolImageHelper.js#L95) makes it look like 0-indexing.
3) This line looks like it accepts both 0 and 1 indexing: https://github.com/beijbom/coralnet/blob/master/project/annotations/static/js/AnnotationToolHelper.js#L966.
4) Point 19 in [this image](https://coralnet.ucsd.edu/image/1103958/annotation/tool/.) actually look like it is outside the image. (This was discovered by @qiminchen in [this PR](https://github.com/beijbom/pyspacer/pull/18)).

**Suggestion**
I'm not a big fan of option 1. It just seems to be inviting a ton of problems down the road. So I think we should go with option 2 or 3. If we do option 2 then we need to clean up and ensure we really do have 1-indexing everywhere in CoralNet. 

@StephenChan : What do you think? How hard is it to do option 3? IMO that would be the most satisfactory way to do it. And it looks like it may help uncover some hidden issues? Btw. I have this feeling we have discussed this before, but I couldn't find a record of it, sorry if I'm repeating myself.

**Comment by beijbom** [comment]
@StephenChan : actually, I was able to transfer this issue over, so I added the tags right here.

**Comment by StephenChan** [comment]
Sounds good.

> It'd be nice to add a bounds verification at the model level -- some "pre-save hook" or something if Django support that. Or even better at the DB layer, but I doubt that is supported since you will need to fetch nrows, ncols from the corresponding image record.

Likely at the model level, yeah. I don't think the DB layer can do anything besides catching values less than 0.

**Comment by StephenChan** [comment]
Another thought: We should roll out this change with some kind of announcement, since it might disrupt workflows that use CSV import/export of annotations. For example, an export done today might not be re-importable next week.

At minimum, a Google Groups post. Perhaps a blog post as well?

**Comment by beijbom** [comment]
Yeah, agreed. A blog-post may be nice. Btw. do you know what CPC uses? I
wouldn't be surprised if that was 1 indexed.

**Comment by StephenChan** [comment]
CPCe is actually 0-indexed, though as you may know, it's also weird since it uses units of 1/15th of a pixel.

Anyway, I did make sure that CoralNet's CPC-upload converted the point locations to 1-indexing (CPC 0 -> CoralNet 1, CPC 49*15 -> CoralNet 50, etc.). In general, it seems that all the currently possible ways of _creating_ Points in CoralNet consistently use 1-indexing. (Which is a relief.)

**Comment by StephenChan** [comment]
All right, after taking a closer look at the CoralNet code, I'd like to clarify the extent of the 0/1 indexing inconsistencies, and discuss whether or not the code has off-by-one errors generally. Worth noting that I have a WIP [branch](https://github.com/beijbom/coralnet/tree/point-rowcol-0-indexing) which addresses a few of these things already.

-----

> it seems some of the consumer code is already 0-indexed, notably
>
> 1. crop-generation [code](https://github.com/beijbom/coralnet/blob/107257fd34cd2c16714b369ec7146ae7222af2c6/project/visualization/utils.py#L129).

```python
    region = im.crop((
        point.column - patch_size // 2,
        point.row - patch_size // 2,
        point.column + patch_size // 2,
        point.row + patch_size // 2
    ))
```

Pillow coordinates do start at (0, 0) at the top left, yes. Interestingly though, this code snippet seems to work equally well for 0 and 1 indexed points.

`im.crop()` acts similarly to Python `range()`, in that the start value is included, while the end value is excluded. So for example, if we have `patch_size` 20 and `point.column` 50 (using 0-indexing), then `region` includes columns 40-59. Not 40-60. Of course, this means the `patch_size` is 20 as intended, but the `point.column` will have 10 pixels to the left of it and 9 pixels to the right of it, thus making it 0.5 pixels off-center.

In general, this code will always create a `region` with even-numbered dimensions, which means there are 4 pixels equally close to the image center. If the point uses 0-indexing, the `point` is the bottom-right of the 4 center pixels. If the point uses 1-indexing, the `point` is the top-left of the 4 center pixels. Either way, `point` ends up 0.5 pixels from the center vertically and horizontally.

That's 0.5 pixels in terms of the original image. The `region`, which is a crop of at least 20% of the image, is then shrunk down to a 150 x 150 patch. So for the image resolutions we generally expect people to upload, those 0.5 pixels are going to be like 0.05-0.2 pixels in the patch. That's probably not going to make a big difference.

If we wanted to try and do better, though, we could change the code to add 1 to both of `region`'s upper bounds, thus always creating a `region` with odd-numbered dimensions, putting the `point` square in the middle. I haven't done that in my WIP branch, but it'd be a simple change to make if we want to do it.

-----

> 2. I'm not an expert on java-script, but since I'd assume java-script uses 0-indexing and I didn't see any `- 1`s I wonder if that one is already using 0-indexing?

Javascript uses 0-indexing, yeah.

> Also, lines like [this](https://github.com/beijbom/coralnet/blob/107257fd34cd2c16714b369ec7146ae7222af2c6/project/annotations/static/js/AnnotationToolImageHelper.js#L95) makes it look like 0-indexing.

Yep, this part deals with setting up the image on the canvas (without worrying about the points yet) and it does assume 0-indexing of the image coordinates.

> 3. This line looks like it accepts both 0 and 1 indexing: https://github.com/beijbom/coralnet/blob/107257fd34cd2c16714b369ec7146ae7222af2c6/project/annotations/static/js/AnnotationToolHelper.js#L966.

```javascript
            var inXRange = imageXMin <= currPoint.column
                           && currPoint.column <= imageXMax;
            var inYRange = imageYMin <= currPoint.row
                           && currPoint.row <= imageYMax;
```

Okay, so this function's variable names are rather confusing. But `imageXMin`, `imageXMax` etc. are the lower and upper bounds of a rectangle selection (start and end of a mouse drag, in 'original image coordinates' rather than display coordinates), and I'm pretty sure both bounds should be inclusive here.

The more interesting Javascript file for this discussion is probably [AnnotationToolHelper.js](https://github.com/beijbom/coralnet/blob/107257fd34cd2c16714b369ec7146ae7222af2c6/project/annotations/static/js/AnnotationToolHelper.js), which is the only JS file that actually deals with the point row/column locations, referred to as `imagePoints` in the file. I traced through the usages of `imagePoints` and the related `canvasPoints`:

- Most parts assume the points are 0-indexed, such as `getNearestPoint()`, which operates the Shift+click functionality of selecting the nearest point to your mouse.

- `pointIsOffscreen()` considers both 0 and `IMAGE_AREA_WIDTH`/`HEIGHT` to be onscreen. Since `canvasPoints` (explained below) is 0-indexed, width/height should be offscreen. So this is just wrong.

- `getCanvasPoints()` assumes `imagePoints` is 1-indexed (which they currently are), and correctly converts them to 0-indexed `canvasPoints`. Basically, there's an HTML canvas element of constant size which displays the point markers on top of the image, and `canvasPoints` tracks the positions of those points in canvas coordinates, given the zoom and pan states of the canvas.

  ```javascript
            // imagePoints[num].row: which image pixel it is, from 1 to height
            // canvasPoints[num].row: offset on the canvas, starting from 0 if the point is visible.
            // Subtract 0.5 so the canvasPoint is in the middle of the point's pixel instead of the bottom-right edge.  Typically won't make much of a difference, but still.
            canvasPoints[imagePoints[i].point_number] = {
                num: imagePoints[i].point_number,
                row: ((imagePoints[i].row - 0.5) * zoomFactor) + imageTopOffset,
                col: ((imagePoints[i].column - 0.5) * zoomFactor) + imageLeftOffset
            };
  ```

  To illustrate what that `- 0.5` does, here's a grid of points on a 6x6 pixel image:
  
  <img src="https://user-images.githubusercontent.com/857711/81272723-b63ce580-9002-11ea-9959-8da055f51552.png" width=500/>

  Note how the point markers are on the centers of the pixels. Here's what it would look like without the `- 0.5`:
  
  <img src="https://user-images.githubusercontent.com/857711/81272746-be952080-9002-11ea-9445-4a19a1a50b67.png" width=500/>

  Obviously 6x6 isn't a typical resolution, but the more you zoom in on a typical image, the more it approaches an issue like this: having the point marker between pixels can throw you off.

  So yeah, I think this part of the point positioning code worked well (it just has to be changed to assume 0-indexed `imagePoints`). However, there's another part which didn't work so well, which leads me to...

-----

> 4. Point 19 in [this image](https://coralnet.ucsd.edu/image/1103958/annotation/tool/.) actually look like it is outside the image. (This was discovered by @qiminchen in [this PR](https://github.com/beijbom/pyspacer/pull/18)).

This is what you're seeing, right?:

![Image 1103958 - No code changes - Zoom 0](https://user-images.githubusercontent.com/857711/81273283-83472180-9003-11ea-9c15-66ced01c5d70.png)

Not sure if you're using the crosshair, but other than that. Anyway, this particular point really has coordinates at the edge of the image (which is 2624 x 2636), so it's expected that part of the 3-pixel-wide crosshair would hang out of the image. But it does look like more than half of the crosshair is hanging out of the image, which is not what we want.

[Here's the culprit](https://github.com/beijbom/coralnet/blob/107257fd34cd2c16714b369ec7146ae7222af2c6/project/annotations/static/js/AnnotationToolHelper.js#L1137), in `drawPointMarker()`:

```javascript
        // Adjust x and y by 0.5 so that straight lines are centered
        // at the halfway point of a pixel, not on a pixel boundary.
        // This ensures that 1-pixel-wide lines are really 1 pixel wide,
        // instead of 2 pixels wide.
        // NOTE: This only applies to odd-width lines.
        x = x+0.5;
        y = y+0.5;
```

Before these statements, `x` and `y` are the correct canvas points. Then we have this 0.5-pixel adjustment for the reason stated in the comment. A couple of reasons why the comment doesn't apply anymore:

- The line width is 3, not 1. Perhaps it was 1 before adding this adjustment, but now it's 3.
- This +0.5 adjustment was committed in 2011 (11c570c44b622ed7fd39b33cccc4007ef0aac662). Firefox and other browsers have been updated since then. In particular, I do not currently see the line width quirk that the comments talks about, even if I change the line width to 1.

The +0.5 is in canvas coordinates (basically, half a screen pixel), so this corresponds to a smaller number of original-image pixels if we're zoomed in.

Here's what it looks like with the +0.5 at various zoom levels:

![Image 1103958 - No code changes](https://user-images.githubusercontent.com/857711/81277943-e936a780-9009-11ea-853f-c56f8565d4cf.png)

And here's what it looks like after removing the +0.5 - the crosshair is slightly farther inside the image in each zoom level, and in particular zoom level 0 looks more centered with the edge:

![Image 1103958 - No +0 5 in drawPointMarker](https://user-images.githubusercontent.com/857711/81278674-d8d2fc80-900a-11ea-81d6-74e2fef77be1.png)

Pretty small difference, all things considered, but we are talking about half a screen pixel after all.

To sum up the second half of this post, I think the only 'wrong' thing about the point positions displayed in the annotation tool was this +0.5 canvas-coords adjustment just before drawing. I actually handled the indexes in a way that calculated the point positions correctly. So in particular, if the user wanted to zoom in a lot to figure out where the center is, the point marker display would have been pretty accurate.

**Comment by beijbom** [comment]
@StephenChan . Thanks for the analysis. Ok, so this is all good to hear. Seems we were mostly consistent, and were we were not, it didn't matter much. 

Regarding the cropping, it seems we should make `patch_size` an odd number and make sure we get the center pixel in the actual center. Although, since this would mean re-generating literally 100's of millions of crops, I'm not so sure it's worth it. What do you think?

**Comment by StephenChan** [comment]
> Regarding the cropping, it seems we should make `patch_size` an odd number and make sure we get the center pixel in the actual center. Although, since this would mean re-generating literally 100's of millions of crops, I'm not so sure it's worth it. What do you think?

I'm not sure if I'm missing a nuance of image-resizing, but I think we just need to make sure `region` has an odd size.

Consider a middling image resolution like 3000 x 2000. When we generate a patch, we first crop out a square `region` which is 20% of the greater dimension, so 600 x 600 pixels. (If we adjust `region` to be odd-sized, then 601 x 601 instead.) Then we resize `region` to the `patch_size`, currently 150 x 150, to get the final patch. Whether `patch_size` is 150 x 150 or 151 x 151, the point's original pixel is going to be squashed into 1/4 of a pixel anyway, so even/odd resolution doesn't seem to matter.

My assumption here is that an image-downsize preserves the 'center of mass' of color values and such, regardless of the initial and final resolutions (as long as they have the same aspect ratio).

**Comment by beijbom** [comment]
Hey. I had forgotten about the 2 stage process. That is good news actually. So we just need to make sure the cropped patch is odd-sized, such that the row, col location is in the center when we crop. Then we can resize to the 
```
region = region.resize((settings.LABELPATCH_NCOLS,
                                        settings.LABELPATCH_NROWS))
```
as previously. It is OK that this is even-sized since the bilinear interpolation will make sure the image _content_ in the middle stays in the middle (now interpolated across 4 pixels).

**Comment by StephenChan** [comment]
All right, thanks for confirming that. I'll just make sure that the crop result is odd-sized then.

**Comment by kriegman** [comment]
Maybe I'm misunderstanding this, but it sounds like the input to the
feature extractor is the usual 224 by 224 but the image is actually cropped
from a larger region.
Is that right? Asking because it raises a question about whether both
qimmin and Tony are cropping images in the same when training the network
and when performing inference offline.

David

**Comment by StephenChan** [comment]
We're talking about the point patches we show to website users, here for example: https://coralnet.ucsd.edu/source/23/browse/patches/ (you can click 'Search' without changing the other fields)

By my understanding, these patches have nothing to do with the feature extractor. Although, if the crops given to the feature extractor are created similarly, then it's worth double checking that cropping code for off-center issues, as well.

**Comment by beijbom** [comment]
Confirming that these has nothing to do with the feature extractor.

**Comment by kriegman** [comment]
That's a relief. Glad I misunderstood.

David

**Comment by StephenChan** [comment]
In terms of coding, this is done in PR #318. Just need to remember the blog post whenever we update production next time. And budget 30 minutes for the DB migration.
